apps/addons/models.py

```diff
@@ -528,6 +523,28 @@ def from_upload(cls, upload, platforms, is_packaged=False, source=None,
                          data.get('default_locale') == addon.default_locale)
         if not locale_is_set:
             addon.default_locale = to_language(translation.get_language())
+
+        return addon
+
+    @classmethod
+    def create_addon_from_upload_data(cls, data, is_listed=True, user=None,
+                                      status=amo.STATUS_NULL):
+        addon = cls.initialize_addon_from_upload(
+            data, is_listed=is_listed, status=status)
+        addon.save()
+        AddonUser(addon=addon, user=user).save()
+        return addon
+
+    @classmethod
+    def from_upload(cls, upload, platforms, is_packaged=False, source=None,
+                    is_listed=True, data=None):
+        from files.utils import parse_addon
+
+        if not data:
+            data = parse_addon(upload)
+
+        addon = cls.initialize_addon_from_upload(
+            is_listed=is_listed, data=data)
```

Comment:
This was the original API that I pulled the guts out of so I think I'd rather leave this one explicit.